zerver/lib/push_notifications.py

```diff
@@ -1307,6 +1353,179 @@ def handle_remove_push_notification(user_profile_id: int, message_ids: list[int]
+        user_profile.realm,
+        COUNT_STATS["mobile_pushes_sent::day"],
+        None,
+        timezone_now(),
+        increment=apple_successfully_sent_count + android_successfully_sent_count,
+    )
+
+
+class RealmPushStatusDict(TypedDict):
+    can_push: bool
+    expected_end_timestamp: int | None
+
+
+class SendNotificationResponseData(TypedDict):
+    android_successfully_sent_count: int
+    apple_successfully_sent_count: int
+    delete_device_ids: list[int]
+    realm_push_status: NotRequired[RealmPushStatusDict]
+
+
+def send_push_notifications(
+    user_profile: UserProfile,
+    apns_payload_data_to_encrypt: dict[str, Any],
+    fcm_payload_data_to_encrypt: dict[str, Any],
+) -> None:
+    # Uses 'zerver_pushdevice_user_bouncer_device_id_idx' index.
+    push_devices = PushDevice.objects.filter(user=user_profile, bouncer_device_id__isnull=False)
+
+    if len(push_devices) == 0:
+        logger.info(
+            "Skipping E2EE push notifications for user %s because there are no registered devices",
+            user_profile.id,
+        )
+        return
+
+    # Prepare payload with encrypted data to send.
+    device_id_to_encrypted_data: dict[str, str] = {}
+    for push_device in push_devices:
+        public_key_str: str = push_device.push_public_key
+        public_key = PublicKey(public_key_str.encode("utf-8"), Base64Encoder)
+        sealed_box = SealedBox(public_key)
+
+        if push_device.token_kind == PushDevice.TokenKind.APNS:
+            encrypted_data_bytes = sealed_box.encrypt(
+                orjson.dumps(apns_payload_data_to_encrypt), Base64Encoder
+            )
+        else:
+            encrypted_data_bytes = sealed_box.encrypt(
+                orjson.dumps(fcm_payload_data_to_encrypt), Base64Encoder
+            )
+
+        encrypted_data = encrypted_data_bytes.decode("utf-8")
+        assert push_device.bouncer_device_id is not None  # for mypy
+        device_id_to_encrypted_data[str(push_device.bouncer_device_id)] = encrypted_data
+
+    # Send push notification
+    try:
+        if settings.ZILENCER_ENABLED:
+            from zilencer.lib.push_notifications import send_e2ee_push_notifications
+
+            response_data: SendNotificationResponseData = send_e2ee_push_notifications(
+                device_id_to_encrypted_data, realm=user_profile.realm
+            )
+        else:
+            post_data = {
+                "realm_uuid": str(user_profile.realm.uuid),
+                "device_id_to_encrypted_data": device_id_to_encrypted_data,
+            }
+            result = send_json_to_push_bouncer("POST", "push/e2ee/notify", post_data)
+            assert isinstance(result["android_successfully_sent_count"], int)  # for mypy
+            assert isinstance(result["apple_successfully_sent_count"], int)  # for mypy
+            assert isinstance(result["delete_device_ids"], list)  # for mypy
+            assert isinstance(result["realm_push_status"], dict)  # for mypy
```

Comment:
This parsing would be a nice place to use Pydantic, to precisely verify the returned objects. Though we probably should not enforce that no extra fields are present when defining that type.

Reply:
#35832